index.rst

```diff
@@ -556,6 +556,67 @@ may not serve in place of monkey patching the file: this caching feature may
 need to be optional (if the temporary file storage is getting too big), and it
 is possible that objects could still exist after the SwiftStorage is destroyed.
 
+Formatter
+---------
+
+A Formatter is an object that has the ability to serialize and deserialize
+objects to and from a type of persistent storage. ``StorageInterface``
+subclasses should use formatters to serialize & deserialize objects.
+
+A Formatter is registered for a class type with the ``StorageInterface``
+subclass that should use the formatter to serialize and deserialize instances of
+that class.
+
+The API of the Formatter is only dependent on the ``StorageInterface`` subclass
+that will be calling it. The signature of functions called by the
+``StorageInterface`` subclass should depend on what parameters are required to
+read from and write to the persistent storage type.
+
+Registering Formatters
+^^^^^^^^^^^^^^^^^^^^^^
+
+Formatters should be registered by calling
+
+.. code-block:: python
+
+    StorageInterface.registerFormatter(formatable, readFormatter, writeFormatter)
+
+Typically ``formatable`` is the class type to serialize & deserialize.
+``readFormatter`` is the function to call when reading and ``writeFormatter`` is
+the function to call when writing.
+
+``StorageInterface`` subclasses may choose to support something other than the
+formatable python class type for the ``formatable`` argument. For example,
+``PosixStorage`` accepts strings as they are specified in the ``storage``
+parameter of the ``datasetType``.
+
+An example of a formatter registration in ``daf_persistence``, in
+``fmtPosixRepositoryCfg.py``:
+
+.. code-block:: python
+
+    PosixStorage.registerFormatter(RepositoryCfg, RepositoryCfgPosixFormatter)
+
+This registers the ``RepositoryCfgPosixFormatter`` as a Formatter for
+``RepositoryCfg`` classes to be written and read via a ``PosixStorage``.
+
+Getting a Formatter
+^^^^^^^^^^^^^^^^^^^
+
+A ``StorageInterface`` subclass can get a Formatter by calling
+``self._getFormatter`` and passing in the class type that will be read or
+written.
```

Comment:
If there are separate read/write formatters do you also need to pass "direction" (either serialize or deserialize)?